### PR TITLE
Fix: point to correct problem part of code, update test

### LINKED ISSUE #2422: clippy prints suggestions in a different style than the source material
Functionally this is not an issue it is just me pointing out a papercut.

My code is as follows:

```rust
impl From<walkdir::Error> for Error {
    fn from(e: walkdir::Error) -> Self {
        let path = e.path()
            .map(|ref_path| ref_path.to_string_lossy().to_string())
            .unwrap_or(String::from("<unknown path>"));

        Error::PathNotFound { path }
    }
}
```

clippy helpfully suggests I use `unwrap_or_else`:

```rust
help: try this
    |
165 |         let path = e.path()
166 |             .map(|ref_path| ref_path.to_string_lossy().to_string()).unwrap_or_else(|| String::from("<unknown path>"));
    |
```

You can see that the `.unwrap_or_else` was printed on the same line as my `.map` instead of the following line.

---

Version information:

```
$ cargo -V
cargo 0.24.0 (45043115c 2017-12-05)

$ rustc -V
rustc 1.23.0 (766bd11c8 2018-01-01)

$ cargo +nightly clippy -- -V
0.0.184
```

### PR DESCRIPTION
**Pull request by bootandy**
Fix span so it no longer contains the whole train-wreck of code and only
points to the problem function (for the unwrap_or lint).

https://github.com/rust-lang-nursery/rust-clippy/issues/2422

Update ui test methods - it had several cases where the error message
span is now shorter

fixes #2422

### CODE REVIEW

File: clippy_lints/src/methods.rs [hunk 1]
```diff
@@ -913,6 +913,10 @@ fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, name: &str, args: &[hir:
             return;
         }
 
+        let start_point = self_expr.span.hi();
+        let end_point = span.hi();
+        let span_replace_word = Span::new(start_point, end_point, span.ctxt());
```
Comment (oli-obk) [inline clippy_lints/src/methods.rs]: You can use `self_expr.span.end_point().to(span.hi())` I think. check the methods on `syntax::code_map::Span`

File: tests/ui/unwrap_or.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#![warn(clippy)]
+
+fn main() {
+    let s = Some(String::from("test string")).unwrap_or("Fail".to_string()).len();
```
Comment (oli-obk) [inline tests/ui/unwrap_or.rs]: can you create a test with the newlines like in the issue?

Comment (bootandy) [reply tests/ui/unwrap_or.rs]: Ah its newlines you want tested? - I thought it was just a method call after the part to replace. I'll add a new test.

Comment (oli-obk) [reply tests/ui/unwrap_or.rs]: You can add it to the same test

File: clippy_lints/src/methods.rs (hunk 1 above)
Comment (bootandy) [reply clippy_lints/src/methods.rs]: Frustratingly thats the wrong Span object. It uses this one: http://manishearth.github.io/rust-internals-docs/syntax_pos/struct.Span.html - took me a while to realise.

However I have reduced it down to: 
`self_expr.span.with_lo(span.hi());`

File: tests/ui/unwrap_or.stderr [hunk 3]
```diff
@@ -6,5 +6,13 @@ error: use of `unwrap_or` followed by a function call
...
   = note: `-D or-fun-call` implied by `-D warnings`
 
-error: aborting due to previous error
+error: use of `unwrap_or` followed by a function call
+ --> $DIR/unwrap_or.rs:8:46
+  |
+8 |       let s = Some(String::from("test string"))
+  |  ______________________________________________^
+9 | |         .unwrap_or("Fail".to_string())
+  | |______________________________________^ help: try this: `.unwrap_or_else(|| "Fail".to_string())`
```
Comment (bootandy) [inline tests/ui/unwrap_or.stderr]: I think this output needs to be double checked. 

Looking at it now.

Comment (bootandy) [reply tests/ui/unwrap_or.stderr]: If we apply this with rustfix then we lose the newline - but the old code does that too so this isn't a regression.

Comment (oli-obk) [reply tests/ui/unwrap_or.stderr]: hmm... I think that was the entire point of the opened issue...

We could create a multi_sugg, one for replacing `unwrap_or` with `unwrap_or_else` and one for replacing `"Fail".to_string()` with `|| "Fail".to_string()`

Comment (bootandy) [reply tests/ui/unwrap_or.stderr]: Yes i think you are right. I got so sidetracked with moving the '^'s to only point at the part of the code that needs changing.

I'll have a play with it and see if I can come up with something.

Comment (bootandy) [reply tests/ui/unwrap_or.stderr]: Pushed a new version. 
Thanks for taking the time to nudge me through this. 
Can we squash all these commits when we merge? I'm conscious that I have gone back and forward quite a bit.

File: clippy_lints/src/methods.rs [hunk 4]
```diff
@@ -845,7 +845,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
 }
 
 /// Checks for the `OR_FUN_CALL` lint.
-fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, name: &str, args: &[hir::Expr]) {
+fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: &Span, name: &str, args: &[hir::Expr]) {
```
Comment (oli-obk) [inline clippy_lints/src/methods.rs]: Please don't use `&Span`, just use `Span`, the type is `Copy`

Comment (bootandy) [reply clippy_lints/src/methods.rs]: Understood, thanks, fixed.

**Comment by oli-obk** [comment]
Thanks!

**Pull request merged**